windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -753,18 +754,33 @@ procedure DownloadingState.Enter;
 var
   DownloadResult: Boolean;
   RetryCount: Integer;
+  FMutex: TKeymanMutex;
 begin
   // Enter DownloadingState
   bucStateContext.SetRegistryState(usDownloading);
 
   RetryCount := 0;
   DownloadResult := False;
+  FMutex := nil;
 
-  while (not DownloadResult) and (RetryCount < 3) do
-  begin
-    DownloadResult := DownloadUpdatesBackground;
-    if not DownloadResult then
-      Inc(RetryCount);
+  try
+    FMutex := TKeymanMutex.Create('KeymanDownloading');
+    // Should be impossible but just exit anyway and let the process current
+    // downloading process finish.
+    if not FMutex.MutexOwned then
```

Comment:
Note, this is a terribly named function (my fault, probably 15 years ago). `MutexOwned` attempts to take ownership of the mutex and then reports whether it was able to. I think we should rename it to `TakeOwnership` (which returns True if it succeeds). `TKeymanMutex` constructor creates the mutex but does not own it.